browser/src/app/Events.ts

```diff
@@ -0,0 +1,337 @@
+ * handle custom events.
+ */
+
+interface EventBaseType {
+	type: string;
+	target: any;
+	[name: string]: any;
+}
+
+type CEventListener = (e: EventBaseType) => void;
+
+interface EventMapObject {
+	[name: string]: CEventListener;
+}
+
+interface CallbackWithContext {
+	fn: CEventListener;
+	ctx: any;
+}
+
+interface CallbackWithoutContext {
+	fn: CEventListener;
+}
+
+class Evented extends BaseClass {
+	private _events: Map<string, Map<string, CallbackWithContext>>;
+	private _numEvents: Map<string, number>;
+
+	private _eventsNoContext: Map<string, Array<CallbackWithoutContext>>;
+	private _eventParents: Map<number, Evented>;
+
+	// This points to the outer L.Evented object.
+	// This is needed till we completely get rid of L.Evented.
+	private _outerObject: any;
+
+	constructor(outerObject?: any) {
+		super();
+		this._events = new Map();
+		this._numEvents = new Map();
+		this._eventsNoContext = new Map();
+		this._eventParents = new Map();
+		this._outerObject = outerObject ? outerObject : this;
+	}
+
+	public on(
+		types: Map<string, CEventListener> | string | EventMapObject,
+		fn: CEventListener | any,
+		context?: any,
+	): Evented {
+		if (typeof types === 'string') {
+			// types can be a string of space-separated words
+			const parts: string[] = Util.splitWords(types);
+
+			for (var i = 0, len = parts.length; i < len; i++) {
+				this._on(parts[i], fn, context);
+			}
+		}
+		// types can be a map of types/handlers
+		else if (types instanceof Map) {
+			for (const [type, listener] of types) {
+				// we don't process space-separated events here for performance;
+				// it's a hot path since Layer uses the on(obj) syntax
+				this._on(type, listener, fn);
+			}
+		} else {
+			const typeNames = Object.keys(types);
+			for (let i = 0, len = typeNames.length; i < len; ++i) {
+				const typeName = typeNames[i];
+				const listener = types[typeName];
+				this._on(typeName, listener, fn);
+			}
+		}
+		return this;
+	}
+
+	public off(
+		types?: Map<string, CEventListener> | string | EventMapObject,
+		fn?: CEventListener | any,
+		context?: any,
+	): Evented {
+		if (!types) {
+			// clear all listeners if called without arguments
+			this._events.clear();
+			this._eventsNoContext.clear();
+			this._numEvents.clear();
+		} else if (typeof types === 'string') {
+			const parts: string[] = Util.splitWords(types);
+
+			for (var i = 0, len = parts.length; i < len; i++) {
+				this._off(parts[i], fn, context);
+			}
+		} else if (types instanceof Map) {
+			for (const [type, listener] of types) {
+				this._off(type, listener, fn);
+			}
+		} else {
+			const typeNames = Object.keys(types);
+			for (let i = 0, len = typeNames.length; i < len; ++i) {
+				const typeName = typeNames[i];
+				const listener = types[typeName];
+				this._off(typeName, listener, fn);
+			}
+		}
+
+		return this;
+	}
+
+	// attach listener
+	private _on(type: string, fn: CEventListener, context: any): void {
```

Comment:
the comment is a good sign this function is misnamed =) can we have attachListener(...) instead ? =)